common/src/main/java/com/lightstep/tracer/shared/CollectorClientProvider.java

```diff
@@ -0,0 +1,123 @@
+package com.lightstep.tracer.shared;
+
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+
+public abstract class CollectorClientProvider {
```

Comment:
Here is where you borrowed the patterns from the `grpc-java` code right? Could you add in a comment here about where the pattern came from? I'm just trying to get a reference too because as we talked about IRL, the code seems fragile.

Reply:
+1